File: sdk/typescript/src/decorators.ts
/**
 * @trace 装饰器（TypeScript 版）。
 *
 * 与 Python SDK 对等：把函数调用记录为 span。未调用 init() 时是零开销直通。
 *
 * TS 没有 Python 那样的 inspect.iscoroutinefunction，但 async 函数自然返回
 * Promise，用同一个 wrapper 处理同步和异步调用。
 *
 * 用法：
 *   import { trace } from "@ariadne/sdk";
 *   class MyService {
 *     @trace("retrieve", { kind: "rag" })
 *     retrieve(query: string) { ... }
 *   }
 */

import { getClient } from "./client";
import { Span } from "./span";

const MAX_ARG_PREVIEW = 512;

function previewArgs(args: unknown[]): string {
  const parts = args.slice(0, 3).map((a) => {
    try {
      return String(a).slice(0, 120);
    } catch {
      return "[unrepr]";
    }
  });
  return parts.join(", ").slice(0, MAX_ARG_PREVIEW);
}

interface TraceOptions {
  kind?: string;
  captureIo?: boolean;
}

/** 包装一个函数，在其执行前后自动产生 span。 */
export function traced<T extends (...args: unknown[]) => unknown>(
  fn: T,
  name: string,
  opts: TraceOptions = {},
): T {
  const kind = opts.kind ?? "internal";
  const captureIo = opts.captureIo ?? true;

  const wrapped = function (this: unknown, ...args: unknown[]): unknown {
    const client = getClient();
    if (!client) return fn.apply(this, args);

    const span = client.span(name, { kind });
    if (captureIo) span.setInput(previewArgs(args));
    span.enter();

    try {
      const result = fn.apply(this, args);
      if (result instanceof Promise) {
        return result.then(
          (val) => {
            if (captureIo && val !== undefined && val !== null) {
              span.setOutput(String(val));
            }
            span.end();
            return val;
          },
          (err: Error) => {
            span.recordError(err);
            span.end();
            throw err;
          },
        );
      }
      if (captureIo && result !== undefined && result !== null) {
        span.setOutput(String(result));
      }
      span.end();
      return result;
    } catch (err) {
      span.recordError(err as Error);
      span.end();
      throw err;
    }
  } as T;
  return wrapped;
}

/** 方法装饰器工厂。与 Python 的 @trace(name, kind=...) 对等。 */
export function trace(
  name?: string,
  opts: TraceOptions = {},
): MethodDecorator {
  const kind = opts.kind ?? "internal";
  const captureIo = opts.captureIo ?? true;

  return function (
    _target: object,
    propertyKey: string | symbol,
    descriptor: PropertyDescriptor,
  ): PropertyDescriptor {
    if (!descriptor || typeof descriptor.value !== "function") {
      return descriptor;
    }
    const spanName = name ?? String(propertyKey);
    descriptor.value = traced(descriptor.value, spanName, { kind, captureIo });
    return descriptor;
  };
}

export { Span };

File: src/ariadne_sdk/decorators.py
"""@trace 装饰器。

同时支持同步与异步函数：async 函数必须返回协程，用同步 wrapper 包会破坏语义。
"""

from __future__ import annotations

import functools
import inspect
from collections.abc import Awaitable, Callable
from typing import Any, ParamSpec, TypeVar, cast

from ariadne_sdk.client import get_client
from ariadne_sdk.span import Span

P = ParamSpec("P")
R = TypeVar("R")

_MAX_ARG_PREVIEW = 512


def _preview_args(args: tuple[Any, ...], kwargs: dict[str, Any]) -> str:
    """生成参数摘要。只取 repr 前若干字符，避免大对象拖慢业务线程。"""
    parts = [repr(a)[:120] for a in args[:3]]
    parts += [f"{k}={v!r}"[:120] for k, v in list(kwargs.items())[:3]]
    return ", ".join(parts)[:_MAX_ARG_PREVIEW]


def _make_span(name: str, kind: str, args: tuple[Any, ...], kwargs: dict[str, Any],
               capture_io: bool) -> Span | None:
    client = get_client()
    if client is None:
        return None
    span = client.span(name, kind=kind)
    if capture_io:
        span.set_input(_preview_args(args, kwargs))
    return span


def trace(
    name: str | None = None,
    *,
    kind: str = "internal",
    capture_io: bool = True,
) -> Callable[[Callable[P, R]], Callable[P, R]]:
    """把函数调用记录为 span。

    未调用 ariadne_sdk.init() 时装饰器是零开销直通，不会报错 ——
    这让同一份代码在有无 Ariadne 的环境下都能跑。
    """

    def decorator(func: Callable[P, R]) -> Callable[P, R]:
        span_name = name or func.__qualname__

        if inspect.iscoroutinefunction(func):

            @functools.wraps(func)
            async def async_wrapper(*args: P.args, **kwargs: P.kwargs) -> Any:
                span = _make_span(span_name, kind, args, kwargs, capture_io)
                if span is None:
                    return await cast("Callable[P, Awaitable[Any]]", func)(*args, **kwargs)
                with span:
                    result = await cast("Callable[P, Awaitable[Any]]", func)(*args, **kwargs)
                    if capture_io and result is not None:
                        span.set_output(repr(result))
                    return result

            return cast("Callable[P, R]", async_wrapper)

        @functools.wraps(func)
        def sync_wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
            span = _make_span(span_name, kind, args, kwargs, capture_io)
            if span is None:
                return func(*args, **kwargs)
            with span:
                result = func(*args, **kwargs)
                if capture_io and result is not None:
                    span.set_output(repr(result))
                return result

        return sync_wrapper

    return decorator
